### PR TITLE
SMART Notebook import

### PR DESCRIPTION
**Pull request by GavinSmith0123**
Linked to issue #404 .

Here is the work I have done on creating a SMART Notebook importer. It is not complete, but I think it is the minimum that would be useful. Importing text, images, pen strokes and page backgrounds is supported.

Before I continue work on missing features (like supporting shapes like rectangles and circles), it would be very helpful to have some feedback before the code diverges too much from the original, and to avoid creating an incompatible version of OpenBoard.

As well as implementing missing features, I also want to look at building the software for Microsoft Windows.

Some of the changes to the `src/adaptors/UBSvgSubsetAdaptor.cpp` file may affect loading native OpenBoard documents:
* I replaced the handling of the <text> SVG element with new code. The old code used the same code to handle these as to handle text elements output by OpenBoard (which are not output as \<text\> elements, but as \<foreignObject\> elements) , with a comment: "This is for backward compatibility with proto text field prior to version 4.3". Since OpenBoard is only at version 1.5 I assume this refers to one of OpenBoard's predecessor projects (OpenSankoré or UniBoard).
* In `loadScene` in `src/adaptors/UBSvgSubsetAdaptor.cpp`, there was a `currentStroke` variable whose purpose I didn't understand. This led to separate strokes in the source file being joined together as one continuous stroke when a file was saved and reloaded. In these changes, a new `UBGraphicsStroke` is created for each stroke in the source file.
* In the SMART Notebook format, it's possible for an image to be referenced more than once (which will happen if a page is duplicated). To accommodate this I allowed image files to be named with their original names rather than a filename based on a "UUID".

I compiled this under Linux Mint 19.2. I tested with my personal collection of SMART Notebook files (most created with SMART Notebook 11). I can supply test files on request.

### CODE REVIEW

File: OpenBoard.pro
Comment (JBBgameich) [inline OpenBoard.pro:441]: (Disclaimer, I'm not a maintainer)

This is distribution specific, an actual fix needs to check multiple paths + check pkgconfig. I'd be in favor of using CMake for the project, since its better suited for finding external dependencies, but this is not plannend upstream as far as I know.

File: src/adaptors/UBImportSMART.cpp
Comment (JBBgameich) [inline src/adaptors/UBImportSMART.cpp:81]: foreach is old Qt api, I'd suggest to make `formats´ const, and iterate through it using
```
for (const QString &format : formats) {
}
```

File: src/adaptors/UBImportSMART.cpp
Comment (JBBgameich) [inline src/adaptors/UBImportSMART.cpp:211]: Iterating on a non-const container can cause detaches,
```suggestion
    const auto sceneItems = scene->items();
    for (QGraphicsItem* item : sceneItems)
```

File: src/adaptors/UBSvgSubsetAdaptor.cpp
Comment (JBBgameich) [inline src/adaptors/UBSvgSubsetAdaptor.cpp:1706]: Indentation seems off here.

File: src/adaptors/UBSvgSubsetAdaptor.cpp
Comment (JBBgameich) [inline src/adaptors/UBSvgSubsetAdaptor.cpp:3151]:
```suggestion
  const auto splitTrString = trString.split(" ", QString::SkipEmptyParts);
  for (const QString &trStr : splitTrString)
```

File: src/adaptors/UBSvgSubsetAdaptor.cpp
Comment (JBBgameich) [inline src/adaptors/UBSvgSubsetAdaptor.cpp:3154]: dead code

**Review by JBBgameich — CHANGES_REQUESTED** [review]
This seems very useful. I am not a maintainer of this project, but I hope my suggestions can be of some use anyway :)

File: src/adaptors/UBImportSMART.cpp
Comment (JBBgameich) [inline src/adaptors/UBImportSMART.cpp:78]:
```suggestion
    const QStringList formats = supportedExtentions();
```

File: src/adaptors/UBSvgSubsetAdaptor.cpp
Comment (GavinSmith0123) [reply src/adaptors/UBSvgSubsetAdaptor.cpp:3155]: It's there in case those parts of the transformation will be supported in the future. I'll remove the code if the maintainers ask me to.

File: OpenBoard.pro
Comment (GavinSmith0123) [reply OpenBoard.pro:441]: Thanks, do you know if there is a way for me to have that patch in my repository so I can build the program myself but not included in the pull request?

**Comment by GavinSmith0123** [comment]
Thanks for your review @JBBgameich ; I've applied most of your suggestions.

The one part of the code I found confusing was the "strokes group". Did you look at or understand the changes I made?

It appears that as part of "stroke optimization", multiple polygon items can be output for a single stroke. However, when they are persisted, they are combined into a single SVG element. When read in again there will just be one polygon item in the graphics scene. I suspect the whole "strokes group" concept could be abandoned, but this is making even more changes.

This interacts with element grouping which is not done correctly at the moment for SMART import.

Did you try my code with any SMART Notebook files?

File: src/adaptors/UBSvgSubsetAdaptor.cpp
Comment (GavinSmith0123) [reply src/adaptors/UBSvgSubsetAdaptor.cpp:3155]: I should add I copied and pasted this code from the CFF Import code (which doesn't work, as far as I can tell).

File: OpenBoard.pro
Comment (JBBgameich) [reply OpenBoard.pro:441]: I'm facing this problem myself, since I have lots of local patches I need for OpenBoard to properly work on my "too new" system. So far I have worked in a branch containing those patches, and use `git rebase -i origin/${BRANCH}` in the end to throw out the unrelated changes before merging.

**Comment by JBBgameich** [comment]
Sorry, I don't have any SMART Notebook files to test with right now. I'm just trying to reduce the review load for the maintainers a bit by doing the generic Qt review part, since the maintainers usually have very limited amount of time.

File: OpenBoard.pro
Comment (GavinSmith0123) [reply OpenBoard.pro:441]: I attempted this and github detected that I'd deleted the branch while I was trying to move it, and closed this pull request. So I've made a fresh pull request for the new branch. https://github.com/OpenBoard-org/OpenBoard/pull/433/files (Unfortunately there are still extraneous changes there too, but I don't want to have to create yet another pull request to fix it.)

What I did in my repository is to add a branch "build-fixes" with the extraneous changes on it, based on the "dev-qt5.1x" branch. I then used git cherry-pick to get rest of the changes, and made this branch "smart-import-build". Finally, I used "git cherry-pick build-fixes..smart-import-build" on a new "smart-import" branch to extract the commits I wanted in the pull request.

**Pull request closed without merging**
